core/src/main/java/org/apache/calcite/rel/core/Join.java

```diff
@@ -230,6 +237,15 @@ public boolean isSemiJoinDone() {
     return false;
   }
 
+  /**
+   * Returns whether this Join is a semijoin.
+   *
+   * @return true if this is semi but without correlate variables.
+   */
+  public boolean isSemiJoin() {
```

Comment:
I'm not sure it is worth adding a new method in public API just for the SemiJoin. Why not having the same equivalent for every other type of join?

Reply:
Cause the semiJoin decision happens everywhere, it is so special cause it only outputs one side.

Reply:
@danny0405 I know it does not happen right now, but in the (near) future, we will also have Join instances with type=ANTI, which also outputs one side; so if we really need to add a new method, I'd propose it to have a more generic name, e.g. moving the SemiJoinType#returnsJustFirstInput logic into here:
```
  public boolean returnsJustFirstInput() {
    switch (this) {
    case SEMI:
    case ANTI:
      return true;
    default:
      return false;
    }
  }
```

Reply:
@rubenada I think `Join#getJoinType().returnsJustFirstInput()` is ok for the `ANTI` case, the reason i add a method for semiJoin is that there are so many cases we have to make decision if it is a SemiJoin type(for rowType or for cost computation or else), after we add in `ANTI` join, i think we just need to add a new method `isAntiJoin`, i'm not sure if we can unify the logic for `ANTI` join where SemiJoin appears(very probably not), if we can, i agree with you to move the `returnsJustFirstInput()` method into `Join`.

Reply:
Maybe it is convenient to have `isSemiJoin` and `isAntiJoin` auxiliary methods for some scenarios (although I tend to agree with @zabetak about why not having them for the rest of join types); but in same cases (e.g. rowType computation) the logic will be the same for both SEMI and ANTI, so in those cases I think it would be useful to have a single method (e.g. returnsJustFirstInput()) rather than calling `isSemiJoin() || isAntiJoin()`